src/v/cloud_storage/inventory/inv_consumer.cc

```diff
@@ -0,0 +1,259 @@
+            if (!flush_state.hashes.empty()) {
+                co_await flush_ntp_hashes(
+                  ntp,
+                  std::move(flush_state.hashes),
+                  flush_state.next_file_name);
+            }
+        }
+        co_return;
+    }
+
+    // Set up a vector to sort entries. The ntp and next filename are not moved
+    // out of the map, only the hashes are moved out.
+    fragmented_vector<flush_entry> entries;
+    ranges::transform(
+      _ntp_flush_states, std::back_inserter(entries), [](auto& e) {
+          return flush_entry{
+            e.first, std::move(e.second.hashes), e.second.next_file_name};
+      });
+
+    ranges::sort(entries, ranges::greater(), [](const auto& e) {
+        return e.hashes.size();
+    });
+
+    const auto target_size = _max_hash_size_in_memory / 3;
+    vlog(
+      cst_log.trace,
+      "Before flush, size in memory {}, target {}",
+      _total_size,
+      target_size);
+
+    size_t idx = 0;
+    while (idx < entries.size() && _total_size >= target_size) {
+        _total_size -= sizeof(entries[idx].hashes[0])
+                       * entries[idx].hashes.size();
+
+        const auto path = entries[idx].ntp.path();
+
+        // Increment the next filename in the map. This must be done before
+        // moving the ntp. Even if the flush fails we have gaps in file names,
+        // which is fine because the filenames can be arbitrary and the hash
+        // loader will read all files in the NTP dir.
+        _ntp_flush_states[entries[idx].ntp].next_file_name += 1;
+
+        co_await flush_ntp_hashes(
+          std::move(entries[idx].ntp),
+          std::move(entries[idx].hashes),
+          entries[idx].file_name);
+
+        vlog(
+          cst_log.trace,
+          "Flushed ntp {}, size in memory {}, target {}",
+          path,
+          _total_size,
+          target_size);
+        idx += 1;
+    }
+
+    vlog(cst_log.trace, "After flush, size in memory {}", _total_size);
+
+    // Put back in all the hashes which were not flushed.
+    for (; idx < entries.size(); ++idx) {
+        _ntp_flush_states[entries[idx].ntp].hashes = std::move(
+          entries[idx].hashes);
+    }
+}
+
+ss::future<> inventory_consumer::flush_ntp_hashes(
+  model::ntp ntp, fragmented_vector<uint64_t> hashes, uint64_t file_name) {
+    auto ntp_hash_path = _hash_store_path / std::string_view{ntp.path()};
+    const auto ntp_hash_dir = ntp_hash_path.string();
+
+    if (!co_await ss::file_exists(ntp_hash_dir)) {
+        co_await ss::recursive_touch_directory(ntp_hash_dir);
+    }
+
+    ntp_hash_path /= fmt::format("{}", file_name);
+    co_return co_await ss::with_file_close_on_failure(
+      ss::open_file_dma(
+        ntp_hash_path.string(), ss::open_flags::create | ss::open_flags::wo),
+      [hashes = std::move(hashes), this](auto& f) mutable {
+          return write_hashes_to_file(f, std::move(hashes));
+      });
+}
```

Comment:
What considerations, if any, do we need to have w.r.t io and reactor priority?

Reply:
This is a good point which is overlooked right now, with this version afaics we will end up using the default priority, I will look into adding an option for this as a follow up.